boat_simulator/common/sensors.py

```diff
@@ -0,0 +1,153 @@
+from dataclasses import dataclass
+
+from typing import Optional, Any
+from numpy.typing import NDArray
+
+
+from boat_simulator.common.types import Scalar, ScalarOrArray
+
+from boat_simulator.common.generators import (
+    ConstantGenerator,
+    MVGaussianGenerator,
+    GaussianGenerator,
+)
+
+WindSensorGenerators = Optional[MVGaussianGenerator | ConstantGenerator]
+GPSGenerators = Optional[GaussianGenerator | ConstantGenerator]
+
+
+@dataclass
+class Sensor:
+    """Interface for sensors in the Boat Simulation."""
+
+    def update(self, **kwargs):
+        """
+        Update attributes in Sensor using keyword arguments.
+
+        Usage: Sensor.update(attr1=val1, attr2=val2, ...)
+
+        Raises:
+            ValueError: If kwarg is not a defined attribute in Sensor
+        """
+        for attr_name, attr_val in kwargs.items():
+            if attr_name in self.__annotations__:
+                setattr(self, attr_name, attr_val)
+            else:
+                raise ValueError(
+                    f"{attr_name} not a property in {self.__class__.__name__} \
+                    expected one of {self.__annotations__}"
+                )
+
+    def read(self, key: str) -> Any:
+        """
+        Read the value from an attribute in Sensor.
+
+        Args:
+            key (str): Attribute name to read from
+
+        Raises:
+            ValueError: If key is not an a defined attribute in Sensor
+
+        Returns:
+            Any: Value stored in attribute with name supplied in "key" argument
+        """
+        if key in self.__annotations__:
+            return getattr(self, key)
+        else:
+            raise ValueError(
+                f"{key} not a property in {self.__class__.__name__}. \
+                Available keys: {self.__annotations__}"
+            )
+
+
+@dataclass
+class WindSensor(Sensor):
+    """
+    Abstraction for wind sensor.
+
+    # TODO: Add delay functions.
+
+    Properties:
+        wind (ScalarOrArray): Wind x, y components or single value
+        wind_noisemaker (Optional[MVGaussianGenerator | ConstantGenerator]):
+        Noise function to emulate sensor noise in wind data reading
+    """
+
+    wind: ScalarOrArray
+    wind_noisemaker: WindSensorGenerators = None
+
+    @property  # type: ignore
+    def wind(self) -> ScalarOrArray:
+        return (
+            self._wind + self.wind_noisemaker.next()  # type: ignore
+            if self.wind_noisemaker is not None
+            else self._wind
+        )
```

Comment:
What if the error vector returned by the noisemaker doesn't match the wind sensor reading? I can see this problem occurring either upon initialization or when updating the data stored.

Reply:
Do you mean when the error vector is a different dimensionality from the wind sensor reading?

Reply:
Yes, the user could construct a noisemaker that does not have the same dimensionality as the wind vector. A similar scenario can happen for the GPS data because the `ConstantGenerator` class accepts both scalars and arrays.

Reply:
Going to mark this as a TODO and merge it.